File: architecture-skills/object-oriented-programming-paradigm/examples/single-responsibility/single-responsibility.py
"""Single Responsibility Principle in Python.

An Invoice that computes totals, formats itself, AND persists itself
serves three actors (finance, presentation, ops) and has three reasons
to change. The refactor gives each actor its own collaborator; the
Invoice keeps only the data it truly owns.
"""

from __future__ import annotations

from dataclasses import dataclass


# --- Before (god class, sketched): three reasons to change ---
#
# class Invoice:
#     def add_item(self, desc, amount): ...   # domain data
#     def total(self): ...                    # changes when FINANCE changes tax
#     def to_text(self): ...                  # changes when LAYOUT changes
#     def save(self): ...                     # changes when STORAGE changes


# --- After: one collaborator per actor ---


@dataclass(frozen=True)
class LineItem:
    description: str
    amount: float


class Invoice:
    """Domain object: owns the data, knows nothing of tax, text, or storage."""

    def __init__(self, invoice_id: str, customer: str) -> None:
        self.invoice_id = invoice_id
        self.customer = customer
        self.items: list[LineItem] = []

    def add_item(self, description: str, amount: float) -> None:
        self.items.append(LineItem(description, amount))


class InvoiceCalculator:
    """Finance's class: only tax-rule changes touch this file."""

    def __init__(self, tax_rate: float) -> None:
        self._tax_rate = tax_rate

    def total(self, invoice: Invoice) -> float:
        subtotal = sum(item.amount for item in invoice.items)
        return subtotal * (1 + self._tax_rate)


class TextInvoiceFormatter:
    """Presentation's class: only layout changes touch this file."""

    def format(self, invoice: Invoice, total: float) -> str:
        lines = [f"Invoice {invoice.invoice_id} for {invoice.customer}"]
        lines += [f"  {i.description:<12} ${i.amount:>7.2f}" for i in invoice.items]
        lines.append(f"  {'TOTAL':<12} ${total:>7.2f}")
        return "\n".join(lines)


class InMemoryInvoiceRepository:
    """Ops' class: only storage changes (schema, driver) touch this file."""

    def __init__(self) -> None:
        self._rows: dict[str, Invoice] = {}

    def save(self, invoice: Invoice) -> None:
        self._rows[invoice.invoice_id] = invoice

    def count(self) -> int:
        return len(self._rows)


if __name__ == "__main__":
    invoice = Invoice("INV-042", "ACME Corp")
    invoice.add_item("Consulting", 1200.00)
    invoice.add_item("Hosting", 90.00)

    calculator = InvoiceCalculator(tax_rate=0.20)
    formatter = TextInvoiceFormatter()
    repository = InMemoryInvoiceRepository()

    total = calculator.total(invoice)
    print(formatter.format(invoice, total))

    repository.save(invoice)
    print(f"saved {invoice.invoice_id} ({repository.count()} record(s) stored)")

File: architecture-skills/object-oriented-programming-paradigm/examples/single-responsibility/single-responsibility.php
<?php
/**
 * Single Responsibility Principle in PHP.
 *
 * An Invoice that computes totals, formats itself, AND persists itself
 * serves three actors. The refactor gives each actor its own class.
 */

declare(strict_types=1);

final readonly class LineItem
{
    public function __construct(
        public string $description,
        public float $amount,
    ) {}
}

final class Invoice
{
    /** @var list<LineItem> */
    public array $items = [];

    public function __construct(
        public readonly string $id,
        public readonly string $customer,
    ) {}

    public function addItem(string $description, float $amount): void
    {
        $this->items[] = new LineItem($description, $amount);
    }
}

final class InvoiceCalculator
{
    public function __construct(private readonly float $taxRate) {}

    public function total(Invoice $invoice): float
    {
        $subtotal = array_sum(array_map(fn(LineItem $i) => $i->amount, $invoice->items));
        return $subtotal * (1 + $this->taxRate);
    }
}

final class TextInvoiceFormatter
{
    public function format(Invoice $invoice, float $total): string
    {
        $lines = ["Invoice {$invoice->id} for {$invoice->customer}"];
        foreach ($invoice->items as $item) {
            $lines[] = sprintf('  %-12s $%8.2f', $item->description, $item->amount);
        }
        $lines[] = sprintf('  %-12s $%8.2f', 'TOTAL', $total);
        return implode(PHP_EOL, $lines);
    }
}

final class InMemoryInvoiceRepository
{
    /** @var array<string, Invoice> */
    private array $rows = [];

    public function save(Invoice $invoice): void
    {
        $this->rows[$invoice->id] = $invoice;
    }

    public function count(): int
    {
        return count($this->rows);
    }
}

$invoice = new Invoice('INV-042', 'ACME Corp');
$invoice->addItem('Consulting', 1200.00);
$invoice->addItem('Hosting', 90.00);

$calculator = new InvoiceCalculator(0.20);
$formatter = new TextInvoiceFormatter();
$repository = new InMemoryInvoiceRepository();

$total = $calculator->total($invoice);
echo $formatter->format($invoice, $total), PHP_EOL;

$repository->save($invoice);
printf("saved %s (%d record(s) stored)\n", $invoice->id, $repository->count());
